src/Functions/toStartOfInterval.cpp

```diff
@@ -40,156 +47,243 @@ class FunctionToStartOfInterval : public IFunction
-                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of argument of function {}. "
-                    "Should be a date or a date with time", arguments[0].type->getName(), getName());
-            first_argument_is_date = isDate(arguments[0].type);
+            const DataTypePtr & type_arg1 = arguments[0].type;
+            if (!isDate(type_arg1) && !isDateTime(type_arg1) && !isDateTime64(type_arg1))
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of 1st argument of function {}. "
+                    "Should be a date or a date with time", type_arg1->getName(), getName());
+            value_is_date = isDate(type_arg1);
         };
 
         const DataTypeInterval * interval_type = nullptr;
-        bool result_type_is_date = false;
-        bool result_type_is_datetime = false;
-        bool result_type_is_datetime_64 = false;
-        auto check_interval_argument = [&]
+        enum class ResultType
         {
-            interval_type = checkAndGetDataType<DataTypeInterval>(arguments[1].type.get());
+            Date,
+            DateTime,
+            DateTime64
+        };
+        ResultType result_type;
+        auto check_second_argument = [&]
+        {
+            const DataTypePtr & type_arg2 = arguments[1].type;
+            interval_type = checkAndGetDataType<DataTypeInterval>(type_arg2.get());
             if (!interval_type)
-                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of argument of function {}. "
-                    "Should be an interval of time", arguments[1].type->getName(), getName());
+                throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of 2nd argument of function {}. "
+                    "Should be an interval of time", type_arg2->getName(), getName());
             switch (interval_type->getKind()) // NOLINT(bugprone-switch-missing-default-case)
             {
                 case IntervalKind::Nanosecond:
                 case IntervalKind::Microsecond:
                 case IntervalKind::Millisecond:
-                    result_type_is_datetime_64 = true;
+                    result_type = ResultType::DateTime64;
                     break;
                 case IntervalKind::Second:
                 case IntervalKind::Minute:
                 case IntervalKind::Hour:
                 case IntervalKind::Day:
-                    result_type_is_datetime = true;
+                    result_type = ResultType::DateTime;
                     break;
                 case IntervalKind::Week:
                 case IntervalKind::Month:
                 case IntervalKind::Quarter:
                 case IntervalKind::Year:
-                    result_type_is_date = true;
+                    result_type = ResultType::Date;
                     break;
             }
         };
 
-        auto check_timezone_argument = [&]
+        enum class ThirdArgument
```

Comment:
Instead of `ThirdArgument` and `third_argument`, you might as well use the `Overload` enum in https://github.com/ClickHouse/ClickHouse/pull/57753/files and also make it a private variable. This expresses the intent more directly and `executeImpl()` and the methods it calls will have access to the overload, e.g. to replace `const size_t time_zone_arg_num = (arguments.size() == 2 || (arguments.size() == 3 && third_argument == ThirdArgument::IsTimezone)) ? 2 : 3;` (sorry, I invented that) by `const size_t time_zone_arg_num = (overload == Overload::Default) ? 2 : 3;`